### ISSUE TITLE
Pytree-based Optimizers

### ISSUE THREAD

**Issue by cgarciae** [deepmind/optax#432]
This topic is in my mind every once in a while, it has already been discussed extensively (e.g. https://github.com/deepmind/optax/issues/197#issuecomment-982548377), but I feel it needs new life because it could resolve the last remaining quirks in optax. 

Optax optimizers have well defined API and contrary to neural networks they have clear ways on how to update their state, making them perfectly suitable for pytree/dataclass interfaces. Similar to what @NeilGirdhar has done [here](https://github.com/NeilGirdhar/tjax/blob/main/tjax/_src/gradient/transforms.py), one could express Pytree version of all optimizers by wrapping functional optax with the added **benefits**:

* Optimizer can now pass through jax's function transformation boundaries, main `jit`.
* Hyper-parameters could be updated using immutable API's like `.replace`.
* You could get rid of the optimizer vs `opt_state` separation.

### Example
For this example I'l be using Flax's `PyTreeNode` but any pytree implementation is just as good enough.
```python
class SGD(PyTreeNode):
  learning_rate: ScalarOrSchedule
  momentum: Optional[float] = None
  nesterov: bool = False
  accumulator_dtype: Optional[Any] = field(pytree_node=False, default=None)
  opt_state: Optional[OptState] = None

  @property
  def tx(self):
    return optax.sdg(**{k: v for k, v in vars(self).items() if k != 'opt_state'})

  def init(self: A, params: Params) -> A:
    return self.replace(opt_state=self.tx.init(params))

  def update(
    self: A, updates: Updates, params: Optional[Params] = None
  ) -> Tuple[Updates, A]:
    updates, opt_state = self.tx.update(updates, self.opt_state, params=params)
    return update, self.replace(opt_state=opt_state)
    
# sample usage
tx = SDG(3e-4)
tx = tx.init(params)
updates, tx = tx.update(grads)
params = optax.apply_updates(params, updates)
```
Some details like he fact that `learning_rate` should be static when its passed a schedule could be solved by creating a new static field `learning_rate_schedule` and some additional logic to send the correct unified parameter to the inner optimizer. 

### Proposal
Given that any community shim will probably not succeed, how about a `optax.pytree` namespace (naming suggestions are welcomed) where these shims could officially live a be discussed with the core team?

**Comment by 8bitmp3** [comment]
@mkunesch @rosshemsley @hbq1 let us know what you think 👍

**Comment by NeilGirdhar** [comment]
Just some thoughts:

> JAX function transformations on classes can be a source of confusion for users. Functions are generally easier to reason about under JAX transformations than class methods.

This proposal doesn't change anything for users since the interface is identical except for the four benefits mentioned.  The reasoning that users have to do is *exactly the same*.  If anything, the user reasoning is simpler since sequence interface is not exposed.

I agree that library developers would have to adapt, but doesn't seem relevant.

This issue is really unfortunate because [I suggested](https://github.com/deepmind/optax/issues/1) the improved design in the very first issue.

> It would also be a good idea to try and "break" this design - e.g. what happens when using more esoteric JAX transforms (such as vmap, pmap, pjit, or grad)

Why don't you try breaking it?

I think it would be good to at least [block the sequence](https://github.com/deepmind/optax/issues/197#issuecomment-983696841), which are misuses of the current optax design.  This will make it easier to improve your design in the future.
